aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -992,59 +992,73 @@ Tensor dense_to_sparse_bsc(const Tensor& self, IntArrayRef blocksize) {
       values.device());
 }
 
+Tensor sparse_compressed_clone(const Tensor& self) {
+  // Just returning self doesn't work
+  // RuntimeError: t.use_count() <= 1 INTERNAL ASSERT FAILED at
+  // "../torch/csrc/autograd/autograd_not_implemented_fallback.cpp":152,
+  // please report a bug to PyTorch.
+  const auto layout = self.layout();
+  Tensor compressed_indices, plain_indices;
+  std::tie(compressed_indices, plain_indices) = at::sparse_csr::getCompressedPlainIndices(self);
+  auto values = self.values();
+  return _sparse_compressed_tensor_unsafe(
+      compressed_indices,
+      plain_indices,
+      values,
+      self.sizes(),
+      values.scalar_type(),
+      layout,
+      values.device());
+}
+
+Tensor sparse_compressed_to_flipped(const Tensor& self, const std::string& name) {
+  const auto layout = self.layout();
+  const auto flipped_layout = at::sparse_csr::flip_compressed_layout(layout);
+  const auto sizes = at::DimVector(self.sizes());
+  const auto transposed_sizes = at::DimVector({sizes[1], sizes[0]});
+
+  Tensor compressed_indices, plain_indices;
+  std::tie(compressed_indices, plain_indices) = at::sparse_csr::getCompressedPlainIndices(self);
+  auto values = self.values();
+
+  const auto is_out_int32 = compressed_indices.scalar_type() == ScalarType::Int;
+  const auto is_transposed_indices = layout == at::kSparseCsc || layout == at::kSparseBsc;
+  auto coo_indices = _convert_indices_from_csr_to_coo(
+      compressed_indices, plain_indices, is_out_int32, true);
+  const auto coo_indices_hashed = at::sparse::flatten_indices(coo_indices,
+      is_transposed_indices ? sizes : transposed_sizes);
+  const auto hash_argsort = std::get<1>(coo_indices_hashed.sort());
+  const auto coo_indices_sorted = coo_indices.index_select(1, hash_argsort);
+
+  const auto new_plain_indices = coo_indices_sorted.select(0, 1);
+  const auto new_compressed_indices_coo_format = coo_indices_sorted.select(0, 0);
+  const auto new_values = values.index_select(0, hash_argsort);
+  const auto new_compressed_indices = _convert_indices_from_coo_to_csr(
+      new_compressed_indices_coo_format,
+      is_transposed_indices ? sizes[0] : transposed_sizes[0],
+      is_out_int32);
+
+  return _sparse_compressed_tensor_unsafe(
+      new_compressed_indices,
+      new_plain_indices,
+      new_values,
+      sizes,
+      new_values.scalar_type(),
+      flipped_layout,
+      new_values.device());
+}
+
 Tensor sparse_compressed_to_sparse_csr(const Tensor& self) {
   if (self.layout() == kSparseCsc) {
     TORCH_CHECK(
         self.dim() == 2,
         "Expected self to be of dimension 2, but got ",
         self.dim(),
         ".");
```

Comment:
It should be possible to support batch dims. A similar approach to what was done to avoid multiple kernel invocations in the `dense->sparse_compressed` conversions could be used. Fold the batch dims into the compressed axis, apply the transformation you have added above to the now 2d CSC matrix, and then unfold.

[This transformation](https://github.com/pytorch/pytorch/blob/4db1588ca09b4f6328c5bd98701e9c302cb39800/aten/src/ATen/native/TensorConversions.cpp#L77-L121) can be used to reintroduce batching in the compressed indices. The reverse of this operation is required for the way in.

Values and plan indices need only be reshaped.

Reply:
I am working on it right now.

Reply:
I don't think that would make the PR too large, @cpuhrsch any opinion on this?

Reply:
It is a bit tricky as you have to convert to a 1d compressed, then to coo, then you have to modify indices to perform a "transposition" and then reverse it all together. It seems to be more involved as far as I can see it now...

Reply:
@amjames , I have added batch dims support.